### PR TITLE
Fix D3D Forced Anisotropy

### PR DESCRIPTION
**Pull request by EmptyChaos**
The D3D backend is known to force anisotropic filtering on all textures when enabled which results in glitches that don't occur in the OpenGL backend under the same anisotropy level. I tried to trace the difference between the 2 implementations to find what was causing this and it seems to actually be an implementation-defined interaction between the anisotropy level and the texture min/mag filter setting in OpenGL. The [Anisotropy specification](https://www.opengl.org/registry/specs/EXT/texture_filter_anisotropic.txt) says that the min/mag filters have an implementation-defined effect on the anisotropic filter; only the combination of `GL_LINEAR`+`GL_LINEAR_MIPMAP_LINEAR` has a clearly defined behavior as being "best effort". I guessed that nVidia's OpenGL drivers were implicitly disabling anisotropy when using `GL_NEAREST` as the filter so I patched D3D with that assumption to see what would happen and got this:
tospatch
(Left is master; right is with this patch applied. Both are 16x Anisotropic in D3D11. The right image matches OpenGL)

I don't know anything about Flipper so this is just a proof of concept.

[Review on Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/3672)

### CODE REVIEW

**Comment by JMC47** [comment]
Considering that this is a fix to an enhancement and brings it to parity with OpenGL, this should be considered for 5.0.  As long as anisotropic filtering off stays the same, I don't see any issue with at least giving this a good, hard look.

**Comment by degasus** [comment]
Sounds fine, but please do the same for OGL, also is this change has no effect. Likely also D3D12 :D

**Comment by Mullin** [comment]
This should be fixed Broken Sunrays in ZTP when using anisotropic filter.

https://wiki.dolphin-emu.org/images/7/72/TwilightPrincessGC_Sunrays_Broken.jpg

**Comment by BhaaLseN** [comment]
Did we ask fifoci for its opinion yet?

@dolphin-emu-bot rebuild

**Comment by delroth** [comment]
FifoCI doesn't run with anisotropic filtering enabled.

**Comment by EmptyChaos** [comment]
@Mullin I've taken a look at those and I don't think I can do anything about it. The way the game implements them is just not compatible with Anisotropic filtering; you'd need an Action Replay code to modify its implementation. The game sets `TexMode0` to `0x00019A` (ie. `min_filter=4,mag_filter=1` = `LINEAR`+`LINEAR`) on the rays so it *wants* maximum filtering, but the angles on the geometry causes Anisotropic filtering to "correct" the image in an undesirable way giving that misrender. Maybe it's possible to do something in the shaders but I'm not an expert.

@BhaaLseN The non-anisotropic code paths are unchanged from master, FifoCI results should be identical.

**Comment by Mullin** [comment]
@EmptyChaos Okay, so sad :(

File: Source/Core/VideoBackends/D3D/D3DState.cpp [hunk 1]
```diff
@@ -259,7 +259,10 @@ ID3D11SamplerState* StateCache::Get(SamplerState state)
 
 	unsigned int mip = d3dMipFilters[state.min_filter & 3];
 
-	if (state.max_anisotropy > 1)
+	// Only use anisotropic filtering if one or both of the filters are set to Linear.
+	// If both filters are set to Point then using anisotropy is equivalent
+	// to "forced filtering" which will cause visual glitches.
+	if (state.max_anisotropy > 1 && (state.min_filter & 4 || state.mag_filter))
```
Comment (lioncash) [inline Source/Core/VideoBackends/D3D/D3DState.cpp]: `(state.min_filter & 4 || state.mag_filter)` should likely be extracted to its own descriptive template function in VideoCommon, considering it's shared across all three backends (perhaps in a new SamplerCommon.h or something)

**Comment by EmptyChaos** [comment]
Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[Source/Core/VideoBackends/D3D/D3DState.cpp, line 265 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3672#-KBMv9vy2Zs8qGXAlVwY-r1-265:-KBRWgqfLB9AONlrZ3XH:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/23c855d09b763b3f7d9fec9b4c25bc5608b57661/Source/Core/VideoBackends/D3D/D3DState.cpp#L265)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3672)*

**Comment by straightflushin** [comment]
Would this fix the text in Mario Party 4-7 with AF enabled?

**Comment by EmptyChaos** [comment]
@straightflushin Possibly? Does it display correctly in OpenGL w/ Anisotropy (on an nVidia GPU, I don't know about AMD/Intel)? I don't have any Mario Partys to test, but I can take a look at a FIFO Log if you want to upload one.

**Comment by straightflushin** [comment]
Sorry, it looks like that bug only happens now with Force Texture Filtering enabled (on both backends).
Anisotropy is fine on those titles.

File: Source/Core/VideoBackends/OGL/SamplerCache.cpp [hunk 2]
```diff
@@ -134,8 +132,26 @@ void SamplerCache::SetParameters(GLuint sampler_id, const Params& params)
...
+	    !IsBpTexMode0PointFiltering(tm0))
+	{
+		// https://www.opengl.org/registry/specs/EXT/texture_filter_anisotropic.txt
+		// For predictable results on all hardware/drivers, only use one of:
+		//	GL_LINEAR + GL_LINEAR (No Mipmaps [Bilinear])
+		//	GL_LINEAR + GL_LINEAR_MIPMAP_LINEAR (w/ Mipmaps [Trilinear])
+		// Letting the game set other combinations will have varying arbitrary results;
+		// possibly being interpreted as equal to bilinear/trilinear, implicitly
+		// disabling anisotropy, or changing the anisotropic algorithm employed.
+		min_filter = (tm0.min_filter & 3) == TexMode0::TEXF_NONE ? GL_LINEAR : GL_LINEAR_MIPMAP_LINEAR;
```
Comment (EmptyChaos) [inline Source/Core/VideoBackends/OGL/SamplerCache.cpp]: @lioncash This test expression is pulled from the D3D backend ([1](https://github.com/dolphin-emu/dolphin/blob/54c2800bb4a7a5bd7ead7319a28da9269a3f8ed3/Source/Core/VideoBackends/D3D/D3DState.cpp#L260), [2](https://github.com/dolphin-emu/dolphin/blob/54c2800bb4a7a5bd7ead7319a28da9269a3f8ed3/Source/Core/VideoBackends/D3D/D3DState.cpp#L313)) and is identical across all four backends (including the [software backend](https://github.com/dolphin-emu/dolphin/blob/54c2800bb4a7a5bd7ead7319a28da9269a3f8ed3/Source/Core/VideoBackends/Software/TextureSampler.cpp#L72)). Do you want me to refactor and pull it out into a function as well?

Comment (lioncash) [inline Source/Core/VideoBackends/OGL/SamplerCache.cpp]: I don't see why not. SamplerCommon should also have a namespace.

**Comment by delroth** [comment]
@dolphin-emu-bot rebuild

**Comment by EmptyChaos** [comment]
I don't think I have any further changes to make. This is ready to be reviewed for merging.

**Comment by JMC47** [comment]
LGTM in terms of what it fixes.

**Comment by RisingFog** [comment]
Can this PR be rebased to latest master?

**Comment by EmptyChaos** [comment]
Rebased on master.

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by RisingFog** [comment]
Anything left holding this PR up?

**Comment by degasus** [comment]
Reviewed 1 of 4 files at r1, 2 of 6 files at r2, 4 of 8 files at r4, 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/VideoBackends/OGL/SamplerCache.cpp, line 63 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3672#-KDYvp-vVKTX0z-8fP-k:-KDYvp-vVKTX0z-8fP-l:-4052974) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f474daca27bbe012b6dca077fe775a74d877685d/Source/Core/VideoBackends/OGL/SamplerCache.cpp#L63)):*
Is this line still fine?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3672)*

**Comment by EmptyChaos** [comment]
Review status: 2 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/VideoBackends/OGL/SamplerCache.cpp, line 63 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3672#-KDYvp-vVKTX0z-8fP-k:-KDaOY6ADVLv1Ssnr0yT:-2124427060) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f474daca27bbe012b6dca077fe775a74d877685d/Source/Core/VideoBackends/OGL/SamplerCache.cpp#L63)):*
Probably not. `min_filter` should be 6 or 4, but that can produce 5 as well. I'll fix it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3672)*

**Comment by delroth** [comment]
@EmptyChaos any progress?

**Comment by EmptyChaos** [comment]
@delroth I've already made the change. I screwed up the Reviewable by not clicking Publish.

**Comment by tapcio** [comment]
So neobrain and/or MaJoRoesch wasn't correct about this, I reported this issue long time ago : https://bugs.dolphin-emu.org/issues/7724 Thanks EmptyChaos.

**Comment by delroth** [comment]
neobrain is correct, this is in some sense a hack to disable anisotropy in
some situations to avoid unwanted effects.

**Comment by EmptyChaos** [comment]
@delroth Not really. The screenshot I posted at the top actually also occurs with Anisotropic Filtering set to 1x (off) and Forced Texture Filtering set to On. The problem is specifically caused by forcing texture filtering when it is meant to be disabled, it isn't directly related to anisotropic filtering itself. You can cause similar issues in some commercial PC games by forcing Anisotropic Filtering from the Nvidia/Catalyst control panel.

Also, I think "fundamental difference between the APIs" may be overstating things a little.
```
A texture's maximum degree of anisotropy is specified independent
from the texture's minification and magnification filter (as
opposed to being supported as an entirely new filtering mode).
...  Implementations are also
permitted to ignore the minification or magnification filter and
implement the highest quality of anisotropic filtering possible.
```
"Implementations are also permitted to **ignore** ..." means this has always been broken in OpenGL as well.

**Pull request merged**